[clang] Add /Zc:__STDC__ flag to clang-cl

xbjfk opened the pull request:
This commit adds the /Zc:\_\_STDC\_\_ argument from MSVC, which defines \_\_STDC_\_.
This means, alongside stronger feature parity with MSVC, that things that rely on \_\_STDC\_\_, such as autoconf, can work.
Link to MSVC documentation of this flag: https://learn.microsoft.com/en-us/cpp/build/reference/zc-stdc?view=msvc-170

AaronBallman commented on clang/test/Driver/cl-zc.cpp:
```diff
@@ -123,6 +123,10 @@
 
 
+// RUN: %clang_cl /dev/null /E -Xclang -dM 2> /dev/null | FileCheck -match-full-lines %s --check-prefix=STDCOFF
+// RUN: %clang_cl /dev/null /E -Xclang -dM /Zc:__STDC__ 2> /dev/null | FileCheck -match-full-lines %s --check-prefix=STDCON
```
I'd also like to see a test that ensures you can't *disable* the option. e.g., `-fno-ms-define-stdc` should be an error.

And this current test actually shows a bug -- this is a .cpp file, not a .c file; `/Zc:__STDC__` should do nothing/diagnose in C++ mode (https://godbolt.org/z/6W4oaKf9e); I'm on the fence as to whether `-fms-define-stdc` should be allowed in C++ mode or not (I'm inclined to say it should also be unsupported in C++ mode).

xbjfk replied:
Good catch, in this case I guess it's best to create a `cl-zc.c` file?

I think it shouldn't be allowed in clang in C++ mode - after all if someone really wants it they can use `-Xclang -fms-define-stdc` or even `-D__STDC__` manually. My guess would be the primary use case of `-fms-compatibility` with clang (rather than clang-cl) is code whose build system is compiler agnostic but the code itself depends on MSVC behavior - in which case they would not be expecting `__STDC__` to be defined.

AaronBallman reviewed (commented):
Thank you for working on this! Please be sure to add a release note to `clang/docs/ReleaseNotes.rst` so users know about the new functionality.

xbjfk commented:
Okay, I made sure it says `argument unused during compilation: '/Zc:__STDC__'` and never defines __STDC__ in C++ mode, but I have couple of questions:
 - Where should I put the test to make sure `-fno-ms-define-stdc` is not accepted?
 -  !IsCXX feels kind of wrong - is this a justified case for creating `isC` in `clang/lib/Driver/Types.cpp`?
    - now that I think about it, what should the behavior be with Objective-C? I imagine the number of people compiling Objective-C with `-fms-compatibility` is very small
 
Thank you

AaronBallman commented:
> Okay, I made sure it says `argument unused during compilation: '/Zc:__STDC__'` and never defines **STDC** in C++ mode, but I have couple of questions:
> 
>     * Where should I put the test to make sure `-fno-ms-define-stdc` is not accepted?

I think that would make sense in `clang/test/Frontend` to ensure you can't do `-Xclang -fno-ms-define-stdc`, probably in a new test file. (I don't see an existing test that would make sense to augment.)

>     * !IsCXX feels kind of wrong - is this a justified case for creating `isC` in `clang/lib/Driver/Types.cpp`?

I know what you mean about it feeling kind of wrong, but it mirrors how the compiler's language options work. Not C++ mode means C mode, so I would probably not add an `isC()` interface. However, if either @MaskRay and @jansvoboda11 feel differently as code owners of that area, I wouldn't be opposed.

>       * now that I think about it, what should the behavior be with Objective-C? I imagine the number of people compiling Objective-C with `-fms-compatibility` is very small

Objective-C is a pure extension over C, so whatever is allowed in C should generally be allowed in Objective-C. Similarly for Objective-C++ and C++.

AaronBallman reviewed (approved):
LGTM!

MaskRay commented on clang/test/Driver/cl-zc.cpp:
```diff
@@ -123,6 +123,15 @@
 // CHECK-CHAR8_T_: "-fno-char8_t"
 
 
+// RUN: %clang_cl /TC /dev/null /E -Xclang -dM 2> /dev/null | FileCheck -match-full-lines %s --check-prefix=NOSTDC
```
Driver tests usually only do `-###` and check the generated cc1 options. This command runs the preprocessor, which can be moved to clang/test/Preprocessor using `%clang_cc1`

xbjfk replied:
Moved to clang/test/Preprocessor, thanks!

MaskRay commented on clang/test/Driver/ms-define-stdc.c:
```diff
@@ -0,0 +1,11 @@
+// ZCSTDCIGNORED-NOT: #define __STDC__ 1
+// ZCSTDCIGNORED: argument unused during compilation
+
+// RUN: not %clang -Xclang -fno-ms-define-stdc %s 2>&1 | FileCheck %s --check-prefix="NOARG"
```
This tests that -fno-ms-define-stdc is not a cc1 option. We almost never do it. There could be a value to test whether the driver -fno-ms-define-stdc option is present.

xbjfk replied:
Hmmm, this test was requested by @AaronBallman - what would you do in this case?

AaronBallman replied:
@MaskRay -- see https://github.com/llvm/llvm-project/pull/68690#discussion_r1355620210 as to why I think it's valuable to have this test (but if you think it's not a useful test, I don't insist on it).

MaskRay replied:
Sorry for the late reply. If you want to show the behavior of the driver option -fno-ms-define-stdc, feel free to add it when you think there is sufficient value.

For CC1 options, I believe we perform a negative test: hey, we check that the negative form is not present.
The value testing that seems low.

xbjfk committed:
[clang] Add /Zc:__STDC__ flag to clang-cl

This commit adds the /Zc:__STDC__ argument from MSVC, which defines __STDC__.
This means, alongside stronger feature parity with MSVC,
that things that rely on __STDC__, such as autoconf, can work.

MaxEW707 commented on clang/include/clang/Basic/LangOptions.def:
```diff
@@ -291,6 +291,7 @@ LANGOPT(HIPStdParInterposeAlloc, 1, 0, "Replace allocations / deallocations with
 
 LANGOPT(OpenACC           , 1, 0, "OpenACC Enabled")
 
+LANGOPT(MSVCEnableStdcMacro , 1, 0, "Define __STDC__ with '-fms-compatability'")
```
```suggestion
LANGOPT(MSVCEnableStdcMacro , 1, 0, "Define __STDC__ with '-fms-compatibility'")
```

xbjfk replied:
Good catch!

MaxEW707 commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -8065,6 +8069,9 @@ void Clang::AddClangCLArgs(const ArgList &Args, types::ID InputType,
    CmdArgs.push_back("-fno-wchar");
  }
 
+ if (!types::isCXX(InputType) && Args.hasArg(options::OPT_fms_define_stdc))
+   CmdArgs.push_back("-fms-define-stdc");
```
This can be deleted. Since `/Zc:__STDC__` is aliased to `fms-define-stdc` this is already handled in the code above.
`AddClangCLArgs` is intended to handle cl.exe options that are not aliased.

If you want `-fms-define-stdc` to be usable when in CL mode then you can add `CLOption` to the visibility as I mentioned above in `Options.td`.

My preference is to remove this code and just add `CLOption`.

See `Option::accept` inside `Option.cpp` for where aliased args are turned into the argument they alias.

xbjfk replied:
Sounds good to me, thanks.

MaxEW707 reviewed (approved):
LGTM!

@xbjfk Let us know if you require one of us to commit on your behalf due to not having commit access.

xbjfk commented:
Hi @MaxEW707 
Yes, I will need someone to commit this for me.
Thanks!
